README.md
# Governify Reporter
[![Node.js CI](https://github.com/governify/reporter/workflows/Node.js%20CI/badge.svg?branch=master)](https://github.com/governify/reporter/actions)
[![Coverage Status](https://coveralls.io/repos/github/governify/reporter/badge.svg)](https://coveralls.io/github/governify/reporter)
<a href="https://standardjs.com"><img src="https://img.shields.io/badge/code_style-semistandard-brightgreen.svg" alt="Standard - JavaScript Style Guide"></a>
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)

Please refer to  [Governify docs](https://docs.governify.io) to get the documentation.

### Running the server
To run the server, run:

```
npm start
```

## Copyright notice

Please refer to  [Governify License](https://docs.governify.io/about/license/) to get Copyright and License details.


CHANGELOG.md
## [2.9.5](https://github.com/governify/reporter/compare/v2.9.4...v2.9.5) (2023-11-08)



## [2.9.4](https://github.com/governify/reporter/compare/v2.9.3...v2.9.4) (2023-11-08)



## [2.9.3](https://github.com/governify/reporter/compare/v2.9.2...v2.9.3) (2023-11-08)



## [2.9.2](https://github.com/governify/reporter/compare/v2.9.1...v2.9.2) (2023-11-08)



## [2.9.1](https://github.com/governify/reporter/compare/v2.9.0...v2.9.1) (2023-07-19)


### Bug Fixes

* node 14 compatibility issue ([14212e5](https://github.com/governify/reporter/commit/14212e5d250f2d29f95b584c8a7bbbd80e4ecd38))



# [2.9.0](https://github.com/governify/reporter/compare/v2.8.0...v2.9.0) (2023-07-05)


### Features

* Fixed calculation for periods ([407588c](https://github.com/governify/reporter/commit/407588cd68a72cf8575c3126795f73efff6f3f2a))
* Transparent gauge block ([539d244](https://github.com/governify/reporter/commit/539d244cde93bb38d31de6070d0a008d092b7505))



# [2.8.0](https://github.com/governify/reporter/compare/v2.7.0...v2.8.0) (2023-06-18)


### Features

* added new block "time-graph2-member-notZero" ([3eaa9e6](https://github.com/governify/reporter/commit/3eaa9e637a267fd4329b4c792791cf3dddd87079))



# [2.7.0](https://github.com/governify/reporter/compare/v2.6.4...v2.7.0) (2023-05-30)


### Features

* Added new block for correlation by member ([ebe490d](https://github.com/governify/reporter/commit/ebe490da9500f347eb3f7041995ee461ffe81a28))



## [2.6.4](https://github.com/governify/reporter/compare/v2.6.3...v2.6.4) (2023-04-27)


### Bug Fixes

* Added missing s for https protocol ([32eff62](https://github.com/governify/reporter/commit/32eff62fcd4c96bb210695c721fdfc8b152322f7))



## [2.6.3](https://github.com/governify/reporter/compare/v2.6.2...v2.6.3) (2023-04-25)


### Bug Fixes

* Changed transformation for member block ([7e45516](https://github.com/governify/reporter/commit/7e45516db40a2620e909e77ab654c0837cc1a518))



## [2.6.2](https://github.com/governify/reporter/compare/v2.6.1...v2.6.2) (2023-04-25)


### Bug Fixes

* Changed download states' URL and icon ([cd58ee0](https://github.com/governify/reporter/commit/cd58ee008220bafb1f3ae253d4905e8886c50a6c))



## [2.6.1](https://github.com/governify/reporter/compare/v2.6.0...v2.6.1) (2023-04-21)


### Bug Fixes

* Changed transformation for renameByRegex ([efbb710](https://github.com/governify/reporter/commit/efbb710fb578d27aaf80c63c0b684f70de2945ad))



# [2.6.0](https://github.com/governify/reporter/compare/v2.5.0...v2.6.0) (2023-04-21)


### Features

* Added support for PSG2-2223 ([c5ffe99](https://github.com/governify/reporter/commit/c5ffe993f91f65df5ba5f127156111d4335b00b6))



# [2.5.0](https://github.com/governify/reporter/compare/v2.4.0...v2.5.0) (2023-02-12)


### Features

* Installed newrelic ([d2ccb10](https://github.com/governify/reporter/commit/d2ccb10d6ba36333c576cc9fc6ad6ceac4109ac7))



# [2.4.0](https://github.com/governify/reporter/compare/v2.3.3...v2.4.0) (2023-02-08)


### Features

* Added new block time-graph-count ([5eec89e](https://github.com/governify/reporter/commit/5eec89edcca8f2039f1cfef6db1786f2e0087195))



## [2.3.3](https://github.com/governify/reporter/compare/v2.3.2...v2.3.3) (2023-02-07)


### Bug Fixes

* Error get 'configDashboard' of undefined ([0a4ad60](https://github.com/governify/reporter/commit/0a4ad60b5b24c5dab70917be282638198f0eec3e))
* Removed empty member from member graph ([9eecbba](https://github.com/governify/reporter/commit/9eecbba997d919924236cb952f5ce0c1a4d7077f))



## [2.3.2](https://github.com/governify/reporter/compare/v2.3.1...v2.3.2) (2022-03-28)


### Bug Fixes

* Axis metric new blocks ([bb913e2](https://github.com/governify/reporter/commit/bb913e23f9ab9289b08dd1cb470693914954394d))



## [2.3.1](https://github.com/governify/reporter/compare/v2.3.0...v2.3.1) (2022-03-28)


### Bug Fixes

* block rename ([71eff0e](https://github.com/governify/reporter/commit/71eff0ee2df6c84911e1e7fedf80b7ec2d6f90e8))



# [2.3.0](https://github.com/governify/reporter/compare/v2.2.0...v2.3.0) (2022-03-25)


### Bug Fixes

* Atom typo name ([b298386](https://github.com/governify/reporter/commit/b298386bdb425a32ffa9c92e0cbf12f6c3686731))


### Features

* New atom and block ([3316bbc](https://github.com/governify/reporter/commit/3316bbc3536f4c0cce30d5304c8ec889b30f9e9e))



# [2.2.0](https://github.com/governify/reporter/compare/v2.1.0...v2.2.0) (2021-10-28)


### Bug Fixes

* Added missing catch blocks in controller ([2c7a3ed](https://github.com/governify/reporter/commit/2c7a3ed04fc3a4ce7139d6dbf79c6364f4979684))
* body parser before commons middleware ([54d9b4a](https://github.com/governify/reporter/commit/54d9b4adffccc2d6384cae5be404bbd70d246687))
* cors before commons middleware ([2b72b1b](https://github.com/governify/reporter/commit/2b72b1b2e37b5d5043a685914676757885aa1246))
* Error abort calculation ([8273c9b](https://github.com/governify/reporter/commit/8273c9bbb0d4bbd63a33857bcc0c6facbb2f4f2b))
* lint ([d17633e](https://github.com/governify/reporter/commit/d17633e638349c1842ed0fdabc1897f9b9a716ee))
* Missing async functions ([353906b](https://github.com/governify/reporter/commit/353906b09f64243acf952f734a4ffb5b942e7bdb))
* notZero blocks error ([e8f64a1](https://github.com/governify/reporter/commit/e8f64a11b5bf0c753d0d676f1913bbf4879d908b))
* notZero gauge error ([dc961a0](https://github.com/governify/reporter/commit/dc961a0f7067bf0b51b8876dc1da765dfcf2bdce))
* ReporterController refactor ([be510c5](https://github.com/governify/reporter/commit/be510c5bc73d22ec0b39eaa60c6b35a3ed46e7b1))
* rm collectors from external infrastructure ([2d8c20d](https://github.com/governify/reporter/commit/2d8c20dbe803df2299f003efca502665ceb76883))
* Show last evidences in evidences view. ([3c955ba](https://github.com/governify/reporter/commit/3c955ba4dd2de28468f2153212bfa92a699ee363))
* Updated governify-commons ([45cd792](https://github.com/governify/reporter/commit/45cd7928ba5cc0f2d1bf64ed7cf418208028eddd))



server.js
/*!
governify-project-gauss-reporter 1.0.0, built on: 2018-04-19
Copyright (C) 2018 ISA group
http://www.isa.us.es/
https://github.com/isa-group/governify-project-gauss-reporter

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>. */

'use strict';

/*
 * Put here your dependencies
 */


module.exports = {
  deploy: _deploy,
  undeploy: _undeploy
};

/**
 * statesAgreementGET.
 * @param {Object} configurations configuration object
 * @param {function} callback callback function
 * @alias module:registry.deploy
 * */
function _deploy(configurations, commonsMiddleware, callback) {


  const governify = require('governify-commons');
  const config = governify.configurator.getConfig('main');
  // Add this to the VERY top of the first file loaded in your app
  var apm = require('elastic-apm-node').start({
    // Override service name from package.json
    // Allowed characters: a-z, A-Z, 0-9, -, _, and space
    serviceName: 'Reporter',
    serviceNodeName: 'Reporter',
    captureBody: 'all',
    transactionMaxSpans: -1,
    usePathAsTransactionName: true,
    abortedErrorThreshold: 0,
    distributedTracingOrigins: ['*'],
    active: config.telemetry.enabled,
    server: config.telemetry.server
  });

  const http = require('http'); // Use http if your app will be behind a proxy.
  const https = require('https'); // Use https if your app will not be behind a proxy.
  const bodyParser = require('body-parser');
  const express = require('express');
  const cors = require('cors');
  const helmet = require('helmet');
  const compression = require('compression');
  const fs = require('fs');
  const path = require('path');

  // Self dependencies
  const logger = require('governify-commons').getLogger().tag('server');
  const swaggerUtils = require('./src/backend/utils').swagger;

  const server = null;
  const app = express();
  
  if (config.server.bypassCORS) {
    logger.info("Adding 'Access-Control-Allow-Origin: *' header to every path.");
    app.use(cors());
  }
  
  logger.info("Using '%s' as HTTP body size", config.server.bodySize);
  app.use(
    bodyParser.urlencoded({
      limit: config.server.bodySize,
      extended: 'true'
    })
  );

  app.use(
    bodyParser.json({
      limit: config.server.bodySize,
      type: 'application/json'
    })
  );

  app.use(commonsMiddleware);


  const frontendPath = path.join(__dirname, './src/frontend');
  const serverPort = process.env.PORT || config.server.port;
  const CURRENT_API_VERSION = 'v4';

  app.use(express.static(frontendPath));

  // Default server options

  app.use(compression());

  

  // Configurable server options

  if (config.server.useHelmet) {
    logger.info('Adding Helmet related headers.');
    app.use(helmet());
  }

  if (config.server.httpOptionsOK) {
    app.options('/*', function (req, res) {
      logger.info('Bypassing 405 status put by swagger when no request handler is defined');
      return res.sendStatus(200);
    });
  }

  if (config.server.servePackageInfo) {
    app.use('/api/info', function (req, res) {
      logger.debug("Serving package.json at '%s'", '/api/info');
      res.json(require('./../../package.json'));
    });
  }

  // latest documentation redirection
  app.use('/api/latest/docs', function (req, res) {
    res.redirect('/api/' + CURRENT_API_VERSION + '/docs');
  });
  app.use('/api/latest/api-docs', function (req, res) {
    res.redirect('/api/' + CURRENT_API_VERSION + '/api-docs');
  });

  app.use('/publicInfrastructure', function (req, res) {
    res.json(governify.infrastructure.getServices().external);
  });

  logger.info('Trying to deploy server');
  if (configurations) {
    logger.info('Reading configuration...');
    for (var c in configurations) {
      var prop = configurations[c];
      logger.info('Setting property' + c + ' with value ' + prop);
      config.setProperty(c, prop);
    }
  }

  // list of swagger documents, one for each version of the api.
  var swaggerDocs = [
    swaggerUtils.getSwaggerDoc(4)
  ];
  // initialize swagger middleware for each swagger documents.
  swaggerUtils.initializeMiddleware(app, swaggerDocs, function () {

    if (config.server.listenOnHttps) {
      https.createServer({
        key: fs.readFileSync('certs/privkey.pem'),
        cert: fs.readFileSync('certs/cert.pem')
      }, app).listen(serverPort, function () {
        logger.info('HTTPS_SERVER mode');
        logger.info('Your server is listening on port %d (https://localhost:%d)', serverPort, serverPort);
        logger.info('Swagger-ui is available on https://localhost:%d/api/%s/docs', serverPort, CURRENT_API_VERSION);
      });
    } else {
      http.createServer(app).listen(serverPort, function () {
        logger.info('Your server is listening on port %d (http://localhost:%d)', serverPort, serverPort);
        logger.info('Swagger-ui is available on http://localhost:%d/api/%s/docs', serverPort, CURRENT_API_VERSION);
        if (callback) {
          callback(server);
        }
      });
    }
  });
}

/**
 * _undeploy.
 * @param {function} callback callback function
 * @alias module:registry.undeploy
 * */
function _undeploy(callback) {
  /* db.close(function () {
      server.close(function () {
        logger.info('Server has been closed');
        callback();
      });
    }); */
  callback();
}
